notifications.go

```diff
@@ -0,0 +1,58 @@
+package packngo
+
+import "fmt"
+
+const notificationBasePath = "/notifications"
+
+// Notification struct
+type Notification struct {
+	ID        string `json:"id,omitempty"`
+	Type      string `json:"type,omitempty"`
+	Body      string `json:"body,omitempty"`
+	Severity  string `json:"severity,omitempty"`
+	Read      bool   `json:"read,omitempty"`
+	Context   string `json:"context,omitempty"`
+	CreatedAt string `json:"created_at,omitempty"`
+	UpdatedAt string `json:"updated_at,omitempty"`
+	User      Href   `json:"user,omitempty"`
+	Href      string `json:"href,omitempty"`
+}
+
+type notificationsRoot struct {
+	Notifications []Notification `json:"notifications,omitempty"`
+	Meta          meta           `json:"meta,omitempty"`
+}
+
+// NotificationService interface defines available event functions
+type NotificationService interface {
+	List(*ListOptions) ([]Notification, *Response, error)
```

Comment:
Can you add `Get` and `MarkAsRead` in this PR.

Reply:
@jasmingacic Thanks for taking a look at this PR.  Thus far I've based this code off of the API swagger docs located at https://api.packet.net/api-docs and cross referencing https://www.packet.net/developers/api/.  Both of which say `Returns a collection of the current user’s notification`, so I may have named this PR incorrectly.  I don't see any APIs that would allow you to `Get` a specific notification based on ID or `MarkAsRead`.  Am I working off the wrong spec?  Or are you asking for caching within this library?

Reply:
These specs have a single GET and PUT https://www.packet.net/developers/api/notifications/

As for caching I'm not sure I mentioned it anywhere. Maybe you are referring to this https://github.com/packethost/packngo/blob/master/devices.go#L126-L142 which is handling cases where there are more than 1 page of results.

Reply:
Thanks for the clarification!  Somehow I got my wires crossed and got the docs from the wrong location, I'll add the GET, PUT, and the pagination to this.

Reply:
p.s. It is just the opposite from pagination. :) Sorry for the confusion. In case that more pages are returned the function should make sure that all returned objects are returned in a single collection.

Reply:
@jasmingacic thanks for taking a look at this, I've implemented the Get and MarkAsRead functions as requested.  Although I had some issues validating API functionality when following the documentation at https://www.packet.net/developers/api/notifications/:

```
Get List of Notifications:
jgavinray@raven:~
$ curl -X GET --header 'Accept: application/json' --header 'X-Auth-Token: REDACTED' 'https://api.packet.net/notifications'
{"notifications":[{"id":"ecb2edbf-b3af-4dbc-8316-e356e10e1d42","type":"improvement","title":"Operating Systems are now on GitHub!","read":null,"body":"\u003cp\u003eOne of the core tenets at Packet is consistency and reproducibility - i.e. the \u0026quot;cloud\u0026quot; experience. To that end, we released a major update that impacts how we create, manage, distribute, and deploy our supported operating systems on Packet. We had four main goals that we wanted to achieve in this process:\u003c/p\u003e\r\n\r\n\u003cul\u003e\r\n\t\u003cli\u003e\u003cstrong\u003eCreating and Exposing \u0026#39;Immutability\u0026#39;\u003c/strong\u003e - By default, we ensure that when you deploy a supported operating system on Packet, you\u0026rsquo;re going to be deploying the most up-to-date and secure version that has the latest upstream updates. This is great for most, but sometimes you need to \u0026ldquo;peg\u0026rdquo; your systems to a particular image/kernel version. You can now pass the git commit during provisioning, and you\u0026#39;re guaranteed that that image will be provisioned.\u003c/li\u003e\r\n\t\u003cli\u003e\u003cstrong\u003eAlways In Sync\u003c/strong\u003e - We now have 15 global datacenters, and it\u0026rsquo;s crucial that our operating system images reside in the facility closest to the system being installed. We\u0026rsquo;ve deployed transparent proxy caches that ensure that no matter which facility you\u0026rsquo;re in, you\u0026rsquo;ll get a fast local copy.\u003c/li\u003e\r\n\t\u003cli\u003e\u003cstrong\u003eTransparency\u003c/strong\u003e - Instead of wondering what just changed when we updated our images, you can check the commit messages in git. You also don\u0026rsquo;t need to wonder how we create our CentOS or Ubuntu images anymore. It\u0026rsquo;s all right here. If it\u0026rsquo;s not there, we\u0026rsquo;re not using it.\u003c/li\u003e\r\n\t\u003cli\u003e\u003cstrong\u003eBYO Images\u003c/strong\u003e - During provisioning, we\u0026rsquo;re initializing the public packet images repository on github.com, which means there\u0026rsquo;s no reason that we can\u0026rsquo;t (you guessed it!) initialize your git repository instead. There are a few bits to work out, but expect READMEs and more details on how to create and deploy your own OS images on Packet soon.\u003c/li\u003e\r\n\u003c/ul\u003e\r\n\r\n\u003cp\u003eTo use the immutability feature, check image_tag of the device you\u0026rsquo;re on using our metadata service:\u003c/p\u003e\r\n\r\n\u003cpre\u003e\r\n\u003ccode\u003e# curl -s https://metadata.packet.net/metadata | \\\r\n    jq -r '.operating_system.image_tag'\r\n47248a016ebc102da71f1541923ef40840b2de63\u003c/code\u003e\u003c/pre\u003e\r\n\r\n\u003cp\u003eYou can then pass that in (or the git short-id) when creating devices, appending it to the operating system slug:\u003c/p\u003e\r\n\r\n\u003cpre\u003e\r\n\u003ccode\u003e{\r\n  \"plan\": \"baremtal_0\",\r\n  \"facility\": \"ewr1\",\r\n  \"operating_system\": \"centos_7:47248a016\"\r\n}\r\n\u003c/code\u003e\u003c/pre\u003e\r\n\r\n\u003cp\u003eIf you have any questions, don\u0026#39;t hesitate to \u003ca href=\"http://help@packet.net\"\u003eemail\u003c/a\u003e us!\u0026nbsp;\u003c/p\u003e","link":"https://www.packet.net/resources/changelog/github-operating-systems/","excerpt":"Today we\u0026#39;ve released a major update to how we manage (and allow you to manage) operating systems on Packet. The new approach leverages GitHub for transparency of all Operating System changes, and provides the ability to \u0026quot;pin\u0026quot; operating systems where you want them, as well as push your own images more easily.","public_id":1178,"created_at":"2018-03-24T00:22:27Z","updated_at":"2018-04-27T18:35:50Z","meta":{"category_color":"#6666cc","title_excerpt":"Operating Systems are now on GitHub!","body_excerpt":"Today we\u0026#39;ve released a major update to how we manage (and allow you to manage) operating systems on Packet. The new approach leverages GitHub for transparency of all Operating System changes, and provides the ability to \u0026quot;pin\u0026quot; operating systems where you want them, as well as push your own images more easily.","image":null}},{"id":"78efd4a5-9ca6-4cd7-b99a-bb77fd8605e8","type":"feature","title":"Backend Transfer","read":null,"body":"\u003cp\u003eBackend transfer is one of our newer features that allows your servers to communicate with each other within a single facility, as well as between Packet facilities\u0026mdash; even across the globe\u0026mdash; using private RFC1918 address space. There\u0026rsquo;s no need to assign additional IP space or configure anything additional on the hosts. Existing devices across facilities will be able to communicate with each other without any additional configuration. The only restriction on backend networking is that all of your servers must be within a single project.\u0026nbsp;\u003c/p\u003e\r\n\r\n\u003cp\u003eThis feature ensures that when you\u0026rsquo;re syncing data between sites, you are doing so in a private and secure manner, without needing to worry about establishing VPN tunnels or communicate via the internet.\u0026nbsp;\u003c/p\u003e\r\n\r\n\u003cp\u003eFor further details about how to enable backend transfer on your project we suggest taking a look at our \u003ca href=\"https://help.packet.net/technical/backend-transfer\"\u003edocumentation\u003c/a\u003e. \u0026nbsp;\u003c/p\u003e\r\n\r\n\u003cp\u003eBackend transfer is currently supported in our AMS1, EWR1, and NRT1 core sites. SJC1 will be enabled pending network upgrades, and support for the Edge facilities will follow soon.\u0026nbsp;\u003c/p\u003e\r\n\r\n\u003cp\u003eQuestions or concerns? Please \u003ca href=\"http://help@packet.net\"\u003eemail\u003c/a\u003e us!\u003c/p\u003e","link":"https://www.packet.net/resources/changelog/backend-transfer/","excerpt":"Backend transfer is one of our newer features that allows your servers to communicate with each other within a single facility, as well as between Packet facilities\u0026mdash; even across the globe\u0026mdash; using private RFC1918 address space. Read here for further release notes.\u0026nbsp;","public_id":1177,"created_at":"2018-03-24T00:20:44Z","updated_at":"2018-04-27T18:36:04Z","meta":{"category_color":"#38d338","title_excerpt":"Backend Transfer","body_excerpt":"Backend transfer is one of our newer features that allows your servers to communicate with each other within a single facility, as well as between Packet facilities\u0026mdash; even across the globe\u0026mdash; using private RFC1918 address space. Read here for further release notes.\u0026nbsp;","image":null}},{"id":"27e6a160-0388-4e2d-bf42-3b129e63447a","type":"feature","title":"Global Anycast IPs","read":null,"body":"\u003cp\u003eGlobal Anycast IPs are public IPv4 addresses that are pulled from Packet-owned IP space and announced in all of our core facilities (edge sites coming soon). This allows you to perform Anycast without having to go through the effort of registering your own ASN and buying public IPv4 space yourself.\u003c/p\u003e\r\n\r\n\u003cp\u003eGlobal Anycast\u0026nbsp;IPs work similarly to normal facility-bound Elastic IPs.\u0026nbsp;\u0026nbsp;They can be assigned to servers statically via our portal/API, or they can be announced and moved between servers via Local BGP.\u003c/p\u003e\r\n\r\n\u003cp\u003eFor further details about how to request and set up Global Anycast IPs, pricing, which facilities support Global Anycast IPs, and how to use Global Anycast IPs please reference our \u003ca href=\"https://help.packet.net/technical/networking/global-anycast-ips\"\u003eknowledge base article\u003c/a\u003e.\u0026nbsp;\u003c/p\u003e\r\n\r\n\u003cp\u003eIf you have questions or want to talk further, please \u003ca href=\"http://help@packet.net\"\u003eemail\u003c/a\u003e us!\u003c/p\u003e","link":"https://www.packet.net/resources/changelog/global-anycast-ips/","excerpt":"Global Anycast IPs are public IPv4 addresses that are pulled from Packet-owned IP space and announced in all of our core facilities (edge sites coming soon). This allows you to perform Anycast without having to go through the effort of registering your own ASN and buying public IPv4 space yourself.","public_id":1176,"created_at":"2018-03-24T00:19:23Z","updated_at":"2018-04-27T18:36:41Z","meta":{"category_color":"#38d338","title_excerpt":"Global Anycast IPs","body_excerpt":"Global Anycast IPs are public IPv4 addresses that are pulled from Packet-owned IP space and announced in all of our core facilities (edge sites coming soon). This allows you to perform Anycast without having to go through the effort of registering your own ASN and buying public IPv4 space yourself.","image":null}}]}

Get Single Notification:
jgavinray@raven:~
$ curl -X GET --header 'Accept: application/json' --header 'X-Auth-Token: REDACTED' 'https://api.packet.net/notifications/ecb2edbf-b3af-4dbc-8316-e356e10e1d42'
{"errors":["Not found"]}

Put on Single Notification:
jgavinray@raven:~
$ curl -X PUT --header 'Accept: application/json' --header 'X-Auth-Token: REDACTED' 'https://api.packet.net/notifications/ecb2edbf-b3af-4dbc-8316-e356e10e1d42'
{"id":"ecb2edbf-b3af-4dbc-8316-e356e10e1d42","read":true}
```
As you can see above, the get calls in raw curl do not return the notification meta data for the notifications list.  This unfortunately will prevent me from addressing returning a single collection of notifications as requested above.  Also when attempting to do a raw curl call and get a specific notification, the API returns an error.  The following curl call to mark as read though is successful....

I reached out to the packet folks in the community slack channel for help in trying to understand what I was doing wrong.  `emiliano.packet` responded "hey @jgavinray notifications is a WIP, I wouldn't work on that feature yet".  So it looks like its a known issue and there is no way to validate the Get functionality at all at this point, so this change is likely dead in the water unfortunately.  Thanks though for the feedback and guidance to set this merged.